Framework/CurveFitting/src/Algorithms/EstimateFitParameters.cpp

```diff
@@ -0,0 +1,381 @@
+  // Initialise the normal distribution parameters (mean and sigma for each
+  // function parameter).
+  std::vector<std::pair<double, double>> distributionParams;
+  for (auto &range : ranges) {
+    auto mean = (range.first + range.second) / 2;
+    auto sigma = std::fabs(range.first - range.second) / 2;
+    distributionParams.push_back(std::make_pair(mean, sigma));
+  }
+
+  auto nParams = costFunction.nParams();
+  Kernel::NormalDistribution distribution;
+  if (seed != 0) {
+    distribution.setSeed(seed);
+  }
+  // Sets of function parameters (GSLVector) and corresponding values of the
+  // cost function (double)
+  using ParameterSet = std::pair<double, GSLVector>;
+  std::vector<ParameterSet> sampleSets(nSamples);
+  // Function for comparing parameter sets.
+  auto compareSets = [](const ParameterSet &p1, const ParameterSet &p2) {
+    return p1.first < p2.first;
+  };
+
+  std::vector<double> badness(nParams);
+  // Run nIterations of the algorithm
+  for (size_t it = 0; it < nIterations; ++it) {
+    for (size_t isam = 0; isam < nSamples; ++isam) {
+      // Draw a set of function parameters from their distributions
+      auto &paramSet = sampleSets[isam];
+      if (paramSet.second.size() != nParams) {
+        paramSet.second.resize(nParams);
+      }
+      for (size_t i = 0; i < nParams; ++i) {
+        auto mean = distributionParams[i].first;
+        auto sigma = distributionParams[i].second;
+        auto p = distribution.randomValue(mean, sigma);
+        paramSet.second[i] = p;
+      }
+      // Calculate the cost function with those parameters
+      costFunction.setParameters(paramSet.second);
+      paramSet.first = costFunction.val() + getConstraints(constraints);
+      if (costFunction.nParams() != nParams) {
+        throw std::runtime_error("Cost function changed number of parameters " +
+                                 std::to_string(nParams) + " -> " +
+                                 std::to_string(costFunction.nParams()));
+      }
+    }
+    // Partially sort the parameter sets in ascending order of the cost
+    // function.
+    // Find nSelection smallest values.
+    std::partial_sort(sampleSets.begin(), sampleSets.begin() + nSelection,
+                      sampleSets.end(), compareSets);
+    // Estimate new distribution parameters from the sample of nSelection sets.
+    GSLVector means(nParams);
+    GSLVector variances(nParams);
+    for (size_t isam = 0; isam < nSelection; ++isam) {
+      auto &paramSet = sampleSets[isam];
+      for (size_t i = 0; i < nParams; ++i) {
+        auto p = paramSet.second[i];
+        means[i] += p;
+        variances[i] += p * p;
+      }
+    }
+    means *= 1.0 / double(nSelection);
+    variances *= 1.0 / double(nSelection);
+    for (size_t i = 0; i < nParams; ++i) {
+      auto mean = means[i];
+      auto sigma = sqrt(variances[i] - mean * mean);
+      if (sigma >= distributionParams[i].second) {
+        badness[i] += double(it + 1);
```

Comment:
What is `badness` used for? I guess it's an indication that the parameter distribution is not converging... but it doesn't seem to be used anywhere else in the code.